packages/transforms/src/buttonElementToComponent/validateCodemodTarget.ts

```diff
@@ -0,0 +1,46 @@
+import { JsxAttribute, Node, StringLiteral } from 'ts-morph'
+
+import { throwFromMethodsIfUndefinedReturn } from '@sourcegraph/codemod-common'
+import { getClassNameRegExp, getParentUntilOrThrow, JsxTagElement } from '@sourcegraph/codemod-toolkit-ts'
+
+import { buttonClassNamesMapping, ClassNameMapping } from './buttonClassNamesMapping'
+
+interface StringLiteralValidatorResult {
+    stringLiteral: StringLiteral
+    classNameMappings: ClassNameMapping[]
+    jsxAttribute: JsxAttribute
+}
+
+export const DEFAULT_ELEMENT_TO_CONVERT = 'button'
+
+// Used in corresponding eslint-rule.
+export const validateCodemodTarget = {
+    /**
+     * Returns `JsxTagElement` is tag name matches, 'button' by default.
+     */
+    JsxTagElement(jsxTagElement: JsxTagElement, tagName = DEFAULT_ELEMENT_TO_CONVERT): JsxTagElement | void {
+        if (jsxTagElement.getTagNameNode().getText() === tagName) {
+            return jsxTagElement
+        }
+    },
+    /**
+     * Returns non-void result if received `StringLiteral`:
+     * 1. Has one of button classes like `btn-primary`.
+     * 2. Has `JsxAttribute` ancestor with `classname` in the attribute name.
+     */
+    StringLiteral(stringLiteral: StringLiteral): StringLiteralValidatorResult | void {
+        const classNameMappings = buttonClassNamesMapping.filter(({ className }) => {
+            return getClassNameRegExp(className).test(stringLiteral.getLiteralValue())
+        })
+
+        if (classNameMappings.length !== 0) {
+            const jsxAttribute = getParentUntilOrThrow(stringLiteral, Node.isJsxAttribute)
```

Comment:
Good point. I thought about that and tried to integrate both [esquery](https://github.com/estools/esquery) and [tsquery](https://github.com/phenomnomnominal/tsquery), which would allow using `JsxTagElement StringLiteral` as a visitor key. And a good thing — it [works in ESLint already](https://eslint.org/docs/developer-guide/selectors). I decided to leave it for future work since it requires implementing a compatibility layer between selectors (even though the mapping of AST nodes is already [available](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/typescript-estree/src/ts-estree/estree-to-ts-node-types.ts) in `@typescript-eslint` package).

Reply:
One of the primary purposes of this PR is to show how to reuse codemods logic and Typescript AST provided by @typescript-eslint in our custom rules, so I wasn't digging into performance. However, it's good to have a reference point. I took a quick look into ESLint rule timing. Below is the screenshot after running this rule over the main Sourcegraph repo. The rule is at the top of the slowest rules and eats 6 seconds:

<img width="600" alt="Screen Shot 2022-01-13 at 16 11 36" src="https://user-images.githubusercontent.com/3846380/149789071-942afdad-d709-4f93-ab81-ab271ce6e21b.png">

After going [through quick wins](https://github.com/sourcegraph/codemod/pull/71/commits/44dd187a5c70795d2dbe6d4528c7b1d9de47a56b) to improve performance it is in a much better place now. 136 milliseconds:

<img width="671" alt="Screen Shot 2022-01-14 at 15 00 43" src="https://user-images.githubusercontent.com/3846380/149789384-d9016235-ab1b-4d7e-a85d-9f4060b797b4.png">

Interestingly, [the dynamic creation of regular expressions](https://github.com/sourcegraph/codemod/blob/vb/latest/packages/transforms/src/buttonElementToComponent/validateCodemodTarget.ts#L33) contributed more than 50% to bad perf. Another thing was removing `getParentUntil` from the `StringLiteral` validator in favor of the `JsxAttribute Literal` selector in the ESLint rule. 

There's still room for improvement, and we can look into it before enabling this plugin in the Sourcegraph repo. Over time we can compile a list of best practices that we should follow to keep eslint-rules performant.